File: generate.js
const util = require('util')
const fs = require('fs');
const handlebars = require('handlebars');
const { performance } = require('perf_hooks');

// Config options
const CONFIG = {
  site: './src/site.json',
  templateDir: './src/templates',
  outputDir: './dist',
};

// Promisify
const readFile = util.promisify(fs.readFile);
const writeFile = util.promisify(fs.writeFile);

// Compiles a handlebar template that has already been read in
function processTemplate(source, page, root) {
  const template = handlebars.compile(source);
  const context = {
    ...page,
    root,
  }
  const output = template(context);
  return output;
}

function safeMkDir(path) {
  return new Promise((resolve) => {
    fs.access(path, fs.constants.F_OK, (err) => {
      if (err) {
        fs.mkdir(
          path,
          { recursive: true },
          (err) => {
            if (err) throw err;
            resolve();
          },
        );
      } else {
        resolve();
      }
    });
  });
}

// Use handlebars to generate templates
function compilePage(page, root) {
  const templatePath = CONFIG.templateDir + '/' + page.template + '.handlebars';
  let output;
  let subCompileJobs;
  return readFile(templatePath, 'utf8')
    .then((source) => {
      // Turn the handlebars template into HTML
      output = processTemplate(source, page, root);

      // If this page has children, compile them too
      subCompileJobs = page.children.map(child => compilePage(child, root))

      return safeMkDir(CONFIG.outputDir + page.path);
    }).then(() => writeFile(
      `${CONFIG.outputDir}${page.path}index.html`,
      output,
      'utf8'
    )).then(() => {
      console.log(`${page.path}index.html`);
      return Promise.all(subCompileJobs)
    });
}

function attachPath(page, parentPath) {
  page.path = `${parentPath}${page.slug}/`

  page.children.forEach((childPage) => {
    attachPath(childPage, page.path);
  });
}

// Entry point - Read site json and compile root
const startTime = performance.now();

readFile(CONFIG.site, 'utf8')
  .then((siteJson) => {
    const root = JSON.parse(siteJson);
    attachPath(root, '');
    return compilePage(root, root);
  })
  .then(() => {
    const elapsed = Math.round(performance.now() - startTime);
    console.log(`Finished in ${elapsed}ms`);
  });

File: generate.rb
require 'json'
require 'liquid'
require 'fileutils'

CONFIG = {
  'site' => './src/site.json',
  'templateDir' => './src/templates',
  'outputDir' => './dist'
}

def process_template (template, page, root)
  @template = Liquid::Template.parse(template)
  parameters = {
    'root' => root
  }.merge(page)
  @template.render(parameters)
end

def attach_path (page, parent_path)
  page['path'] = "#{parent_path}#{page['slug']}/"
  page['children'].each do |childPage|
    attach_path(childPage, page['path'])
  end
end

def write_file (output, path)
  dir_name = CONFIG['outputDir'] + path
  unless File.directory?(dir_name)
    FileUtils.mkdir_p(dir_name)
  end
  File.write(dir_name + 'index.html', output)
end

def compile_page (page, root)
  template_path = CONFIG['templateDir'] + '/' + page['template'] + '.liquid';
  template_string = File.read(template_path)
  output = process_template(template_string, page, root)
  write_file(output, page['path'])

  page['children'].each do |childPage|
    compile_page(childPage, root)
  end
end

starting = Process.clock_gettime(Process::CLOCK_MONOTONIC)

siteRaw = File.read(CONFIG['site'])
root = JSON.parse(siteRaw)
withPath = attach_path(root, '')
compile_page(root, root)

ending = Process.clock_gettime(Process::CLOCK_MONOTONIC)
elapsed = (ending - starting) * 1000
puts "Finished in #{elapsed.round(1)} milliseconds"